File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not(root):
            return []
        if (root.left is None) and (root.right is None):
            return [root.val]
        if root.left is None:
            res = [root.val]
            res.extend(self.inorderTraversal(root.right))
            return res
        if root.right is None:
            res = self.inorderTraversal(root.left)
            res.append(root.val)
            return res
        res = self.inorderTraversal(root.left)
        res.append(root.val)
        res.extend(self.inorderTraversal(root.right))
        return res

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        if (root == NULL) {
            vector<int> res;
            return res;
        }
        if ((root->left == NULL) & (root->right == NULL)) {
            vector<int> res;
            res.push_back(root->val);
            return res;
        }
        if (root->left == NULL) {
            vector<int> res;
            res.push_back(root->val);
            vector<int> sub_res = this->inorderTraversal(root->right);
            for (auto &n : sub_res) {
                res.push_back(n);
            }
            return res;
        }
        if (root->right == NULL) {
            vector<int> res = this->inorderTraversal(root->left);
            res.push_back(root->val);
            return res;
        }
        vector<int> res = this->inorderTraversal(root->left);
        res.push_back(root->val);
        vector<int> sub_res = this->inorderTraversal(root->right);
        for (auto &n : sub_res) {
            res.push_back(n);
        }
        return res;
    }
};